src/smt/optimization_solver.h

```diff
@@ -80,38 +85,163 @@ class Objective
+ * A solver for optimization queries.
+ *
+ * This class is responsible for responding to optmization queries. It
+ * spawns a subsolver SmtEngine that captures the parent assertions and
+ * implements a linear optimization loop. Supports activateObjective,
+ * checkOpt, and objectiveGetValue in that order.
+ */
+class OptimizationSolver
+{
+ public:
+  /** parent is the smt_solver that the user added their assertions to **/
+  OptimizationSolver(SmtEngine* parent);
+  ~OptimizationSolver();
+
+  /** Runs the optimization loop for the activated objective **/
+  OptResult checkOpt();
+  /** Activates an objective: will be optimized for
+   * Parameter bvSigned specifies whether we should use signed/unsigned
+   * comparison for BitVectors (only effective for BitVectors)
+   * and its default is false **/
+  void activateObj(const Node& obj,
+                   const int& type,
+                   bool bvSigned=false);
+  /** Gets the value of the optimized objective after checkopt is called **/
+  Node objectiveGetValue();
+
+ private:
+  /** The parent SMT engine **/
+  SmtEngine* d_parent;
+  /** The objectives to optimize for **/
+  Objective d_activatedObjective;
+  /** A saved value of the objective from the last sat call. **/
+  Node d_savedValue;
+};
+
+/**
+ * Optimizer for individual CVC4 type
+ * Currently supported: Integer, BitVector
+ */
+class OMTOptimizer
+{
+ public:
+  virtual ~OMTOptimizer() = default;
+  /** Given a target node, retrieve an optimizer specific for the node's type
+   * the second field isSigned specifies whether we should use signed comparison
+   * for BitVectors and it's only valid when the type is BitVector
+   **/
+  static std::unique_ptr<OMTOptimizer> getOptimizerForNode(
+      Node targetNode, bool isSigned = false);
+  /** Minimize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> minimize(SmtEngine* parentSMTSolver,
+                                              Node target) = 0;
+  /** Maximize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> maximize(SmtEngine* parentSMTSolver,
+                                              Node target) = 0;
+};
+
+/**
+ * Optimizer for Integer type
+ */
+class OMTOptimizerInteger : public OMTOptimizer
+{
+ public:
+  virtual ~OMTOptimizerInteger() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver
```

Comment:
we don't need to repeat the description of the function if it is the same -- enough to do so in the base class only.

Reply:
ok